### PR TITLE
Add logging and rule application trace

### PR DESCRIPTION
**Pull request by Kieranoski702**
This PR adds logging into main and rewrite.rs. The default log level is `INFO` which prints the same information that the print statements in main would print which have now been replaced with calls to the log macro. Using the environment variable `LOG_LEVEL`, you can set the log level to one of `TRACE`, `DEBUG`, `INFO`, `WARN`, and `ERROR`. Nothing makes use of debug currently but trace will output each attempted rule application as well as successful rule applications to a log file called `conjure_oxide.log`. 

@niklasdewally @lixitrixi @ozgurakgun @gskorokhod There are currently a few issues that I would ideally like to solve before this PR is merged that I'd like some help with:

1. I have tried to replicate the nice output style that the prints used within main by using `serde:to_string_pretty` but it seems like the logger does not evaluate escaped characters like `\n` and instead displays them so it does not look "pretty"

2. I have only added logging to main and rewrite (mainly because I am familiar with these modules) but I am sure there are plenty of more places where logging would be useful at multiple different log levels. I would appreciate some suggestions on which modules need logging, what exactly should be logged in these modules and which levels/output methods should be used for these.

3. I am still unsure if all warnings and errors should be replaced with WARN and ERROR level logs as there is a lot of discussion online about which should be used mainly because WARN and ERROR log levels can be turned off and because they are "not standard practice". Some advice on this would be appreciated!

After this is merged, I would encourage everyone to write logging for their future modules. Send me a message if you need help with setting up logging but the log messages in this PR are good examples.

### CODE REVIEW

File: conjure_oxide/src/main.rs [hunk 1]
```diff
@@ -28,32 +31,43 @@ struct Cli {
...
 pub fn main() -> AnyhowResult<()> {
+    let log_file = File::options()
+        .create(true)
+        .append(true)
+        .open("conjure_oxide.log")
+        .unwrap();
+
+    Builder::new()
+        .with_target_writer("info", new_writer(stdout()))
+        .with_target_writer("file", new_writer(log_file))
```
Comment (niklasdewally) [inline conjure_oxide/src/main.rs:42]: Is whether it goes to stout or log_file dependant on the target= variable, or is it like python where (eg) warnings and above go to console by default

**Comment by niklasdewally** [comment]
@Kieranoski702 

Will there be a mechanism in which the logger prints what layer a message comes from? 

For example, it would be nice to call `log::error("found something I don't like!")` inside a SolverAdaptor and it print as `[MinionAdaptor] ERROR: found something I don't like!`. In Java /  Python it usually is the name of the class it comes from but there might be a better approach in Rust.

In particular, for me, it would be good to be able to differentiate between logs created by Solver and those created by a specific SolverAdaptor. 

Some ideas of things that could be logged for the Solver state machine: 

* (INFO) state transitions; for example "loaded model", "solve started", "solve finished". 
* (INFO or TRACE) the status of a finished solve.
* (INFO or TRACE) the status of a finished solve.
* (INFO or TRACE) wall time.
* The user should also be able to add their own log calls within their given callback.


I would be happy to help add some of this to Solver.

File: conjure_oxide/src/main.rs (hunk 1 above)
Comment (Kieranoski702) [reply conjure_oxide/src/main.rs:42]: It's the target= parameter so for each log call you can set the output method, apologies for not clarifying!

Comment (ChrisJefferson) [reply conjure_oxide/src/main.rs:42]: My instinct is that people writing logs shouldn't be choosing between 'info' and 'file' -- they should choose an importance level, and then the importance level can decide where the event goes (screen, file, or nowhere).

I'd be tempted (open to debate however!) to change the target_writer to (I hopefully get the notation right)

.with_target_writer("*", new_writer(log_file))

Then encourage people to use the `target` for the subsystem the warning is coming from, which I have found useful in the past, as sometimes I just want to look at the log from a single subsystem (which you can just do with `grep`/`rg` from the command line, or do something more clever which changes how the log is produced :) )

**Comment by ozgurakgun** [comment]
Sorry @Kieranoski702 I merged something else first! But the conflict looks trivial. Can you please fix the conflict and I can merge this?

**Pull request merged**